### PR TITLE
Convert some std error types to PyErr

### PR DESCRIPTION
**Pull request by messense**
Address #21

### CODE REVIEW

File: src/err.rs [hunk 1]
```diff
@@ -369,6 +370,90 @@ impl std::convert::From<PyErr> for std::io::Error {
     }
 }
 
+impl std::convert::From<std::io::Error> for PyErr {
+    fn from(err: std::io::Error) -> Self {
+        let py = unsafe { Python::assume_gil_acquired() };
+        PyErr::new::<exc::IOError, _>(py, err.description())
```
Comment (fafhrd91) [inline src/err.rs]: This should be more precise, check utils.rs from async-tokio

**Comment by fafhrd91** [comment]
Maybe we can change PyResult type and include unconverted errors. Then we can use custom form of From trait in callback handler when we have python object available. Also we can use unstable `try_traiy` for implementing `?`

**Comment by fafhrd91** [comment]
Another option is to modify PyErr and just store errors unconverted. And convert them inside `PyErr::restore()`

**Comment by messense** [comment]
I known it's unsafe, but without it you would more likely to just `unwrap()` to workaround it which it not good either.

I'd like to look into other options you mentioned to make it safer.

**Comment by messense** [comment]
I added a `ToPyErr` trait just like that in `async-tokio`.

The ideal way would be

```rust
fn my_py_fn(py: Python, path: &str) -> PyResult<String> {
  let f = File::open(path)?;
  ...
}
```

But that's a little complicated to implement now. With this PR we can do:

```rust
use pyo3::ToPyErr;

fn my_py_fn(py: Python, path: &str) -> PyResult<String> {
  let f = File::open(path).map_err(|e| e.to_pyerr(py))?;
  ...
}
```

which should be good enough for most cases.

**Comment by fafhrd91** [comment]
did you look into this feature? https://github.com/rust-lang/rust/pull/42275

**Comment by fafhrd91** [comment]
Lgtm. Could you also update async-tokio

**Pull request merged**